### ISSUE TITLE
[1.0.4] Operator destroys healthy ceph cluster after restart

### ISSUE THREAD

**Issue by mmack** [rook/rook#3575]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
Operator, after a restart / reschedule destroys a healthy cluster

**Expected behavior:**
Should not happen

**How to reproduce it (minimal and precise):**
We tried to deploy a fresh v1.0.4 ceph cluster with our battletested config which is running 3 clusters over a year now. It works all out until the operator restarts. The logs speak for itself, the mons get deleted.

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

 To get logs, use `kubectl -n <namespace> logs <pod name>`
When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
Read [Github documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Environment**:
* OS (e.g. from /etc/os-release):
Coreos
* Kernel (e.g. `uname -a`):
4.14.96-coreos
* Cloud provider or hardware configuration:
Hardware / VMware
* Rook version (use `rook version` inside of a Rook Pod):
v1.0.4
* Storage backend version (e.g. for ceph do `ceph -v`):
ceph version 14.2.1 (d555a9489eb35f84f2e1ef49b77e19da9d113972) nautilus (stable)
* Kubernetes version (use `kubectl version`):
1.13.4
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
systemd deployed

Interesting parts of the log files:
```
2019-08-06 04:47:07.698574 I | op-mon: loaded: maxMonID=4, mons=map[a:0xc0012ebb60], mapping=&{Node:map[d:0xc0012ef5c0 e:0xc0012ef5f0 a:0xc0012ef4d0 b:0xc0012ef560 c:0xc0012ef590] Port:map[]}
2019-08-06 04:47:07.704874 D | op-mon: updating config map rook-ceph-mon-endpoints that already exists
2019-08-06 04:47:07.708128 I | op-mon: saved mon endpoints to config map map[data:a=100.71.255.48:6789 maxMonId:4 mapping:{"node":{"a":{"Name":"k8s-worker-00.lxstage.domain.com","Hostname":"k8s-worker-00.lxstage.domain.com","Address":"172.22.254.105"},"b":{"Name":"k8s-worker-101.lxstage.domain.com","Hostname":"k8s-worker-101.lxstage.domain.com","Address":"172.22.254.183"},"c":{"Name":"k8s-worker-102.lxstage.domain.com","Hostname":"k8s-worker-102.lxstage.domain.com","Address":"172.22.254.186"},"d":{"Name":"k8s-worker-103.lxstage.domain.com","Hostname":"k8s-worker-103.lxstage.domain.com","Address":"172.22.254.185"},"e":{"Name":"k8s-worker-104.lxstage.domain.com","Hostname":"k8s-worker-104.lxstage.domain.com","Address":"172.22.254.187"}},"port":{}}]
```
This indicates that we have 5 running mons. No need to update anything... and later:

```
2019-08-06 04:47:09.796587 I | op-mon: targeting the mon count 5
2019-08-06 04:47:09.914452 D | op-mon: there are 22 nodes available for 1 mons
2019-08-06 04:47:10.108140 D | op-mon: mon pod on node k8s-worker-00.lxstage.domain.com
2019-08-06 04:47:10.108370 I | op-mon: Found 4 running nodes without mons
2019-08-06 04:47:10.108382 D | op-mon: mon a already assigned to a node, no need to assign
2019-08-06 04:47:10.108388 D | op-mon: mon f assigned to node k8s-worker-101.lxstage.domain.com
2019-08-06 04:47:10.108393 D | op-mon: using IP 172.22.254.183 for node k8s-worker-101.lxstage.domain.com
2019-08-06 04:47:10.108400 D | op-mon: mon g assigned to node k8s-worker-102.lxstage.domain.com
2019-08-06 04:47:10.108407 D | op-mon: using IP 172.22.254.186 for node k8s-worker-102.lxstage.domain.com
2019-08-06 04:47:10.108411 D | op-mon: mon h assigned to node k8s-worker-103.lxstage.domain.com
2019-08-06 04:47:10.108416 D | op-mon: using IP 172.22.254.185 for node k8s-worker-103.lxstage.domain.com
2019-08-06 04:47:10.108422 D | op-mon: mon i assigned to node k8s-worker-104.lxstage.domain.com
2019-08-06 04:47:10.108427 D | op-mon: using IP 172.22.254.187 for node k8s-worker-104.lxstage.domain.com
2019-08-06 04:47:10.108436 D | op-mon: mons have been assigned to nodes
2019-08-06 04:47:10.108442 I | op-mon: creating mon f
2019-08-06 04:47:10.108454 D | op-k8sutil: creating service rook-ceph-mon-a
2019-08-06 04:47:10.343842 D | op-k8sutil: updating service %s
2019-08-06 04:47:10.706629 I | op-mon: mon a endpoint are [v2:100.71.255.48:3300,v1:100.71.255.48:6789]
2019-08-06 04:47:10.706661 D | op-k8sutil: creating service rook-ceph-mon-f
2019-08-06 04:47:11.196724 I | op-mon: mon f endpoint are [v2:100.71.47.109:3300,v1:100.71.47.109:6789]
2019-08-06 04:47:11.508132 D | op-mon: updating config map rook-ceph-mon-endpoints that already exists
```

why is it updating anything? I see he want's to add 2 new mons which is nonsense.. we don't have the nodes in our tolerations for that...

cluster config: (It's healthy again after we applied a backup from yesterday)
```
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  annotations:
  creationTimestamp: "2019-08-05T15:05:28Z"
  finalizers:
  - cephcluster.ceph.rook.io
  generation: 56
  name: rook-ceph
  namespace: rook-ceph-stage-primary
  resourceVersion: "307846765"
  selfLink: /apis/ceph.rook.io/v1/namespaces/rook-ceph-stage-primary/cephclusters/rook-ceph
  uid: 76235f05-b792-11e9-9b32-0050568460f6
spec:
  annotations:
    all:
      kubernetes.io/egress-bandwidth: 100000M
      kubernetes.io/ingress-bandwidth: 100000M
  cephVersion:
    image: ceph/ceph:v14.2.1-20190430
  dashboard:
    enabled: true
    port: 7000
    ssl: false
  dataDirHostPath: /opt/rook/rook-ceph-stage-primary
  mon:
    allowMultiplePerNode: false
    count: 5
    preferredCount: 0
  network:
    hostNetwork: false
  placement:
    mgr:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: rook-namespace
              operator: NotIn
              values:
              - rook-ceph-stage-primary
    mon:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: rook-namespace
              operator: In
              values:
              - rook-ceph-stage-primary
          - matchExpressions:
            - key: node-role.kubernetes.io/quorum
              operator: In
              values:
              - ""
      tolerations:
      - effect: PreferNoSchedule
        key: node-role.kubernetes.io/quorum
        operator: Exists
    osd:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: rook-namespace
              operator: In
              values:
              - rook-ceph-stage-primary
      tolerations:
      - effect: PreferNoSchedule
        key: node-role.kubernetes.io/storage
        operator: Exists
  rbdMirroring:
    workers: 0
  resources:
    all:
      limits:
        cpu: 500m
        memory: 1Gi
      requests:
        cpu: 500m
        memory: 1Gi
    mgr:
      limits:
        cpu: "2"
        memory: 1Gi
      requests:
        cpu: "2"
        memory: 1Gi
  storage:
    config:
      databaseSizeMB: "1024"
      journalSizeMB: "1024"
      osdsPerDevice: "6"
      storeType: bluestore
    deviceFilter: ^nvme
    nodes:
    - config: null
      location: datacenter=rz1,rack=r1
      name: k8s-worker-101.lxstage.domain.com
      resources: {}
    - config: null
      location: datacenter=rz2,rack=r2
      name: k8s-worker-102.lxstage.domain.com
      resources: {}
    - config: null
      location: datacenter=rz1,rack=r1
      name: k8s-worker-103.lxstage.domain.com
      resources: {}
    - config: null
      location: datacenter=rz2,rack=r2
      name: k8s-worker-104.lxstage.domain.com
      resources: {}
    useAllDevices: false
status:
  ceph:
    health: HEALTH_OK
    lastChanged: "2019-08-06T06:57:14Z"
    lastChecked: "2019-08-06T07:18:11Z"
    previousHealth: HEALTH_WARN
  state: Created

```

Here's the complete operator log: https://gist.github.com/mmack/1a998cf876dd9638bfe2e0e8a1bd55eb

**Comment by mmack** [comment]
ok it's getting really strange here... don't know how to describe that technically, but i had a running cluster. health OK. Everything working. If delete a mon deployment (for example mon-a) and what happens is that all other deployments are getting deleted also, including all osd deployments. Left is a cephcluster in state "Creating".

The operator log is not helpfull at all: https://gist.github.com/mmack/d1c50f1360000e5c84fdffe5278e6321

i'm kinda lost here... any help appreciated, we can do the same run again with some of you guys watching, if that helps...

**Comment by mmack** [comment]
for the record, here three log files: 
1. multiple operator logs combined (there are 3 restarts or so)
2. `while true; do kubectl get configmap rook-ceph-mon-endpoints -o yaml >> rook-ceph-mon-endpoints.yaml; sleep 1; done; ` : https://gist.github.com/mmack/7b3e8423964e4573b69f47fff4d9ed0a
3. ceph -w while testing: https://gist.github.com/mmack/3dbcfd4bb495144b7738294bdc782f9c

**Comment by mmack** [comment]
We redeployed 1.0.1 and the problem is gone. fyi

**Comment by travisn** [comment]
@mmack Can you clarify what you mean by the cluster being destroyed? All the mon and osd pods are disappearing? I'm not seeing in the logs were that happened, only that the mons are calling "update" and trying to failover at unexpected times. Overall I'm struggling to follow what happened here. 

Update is called as a regular part of the orchestration, even though nothing is actually updated in the pod spec until it is an upgrade or there was some change to the cephcluster CR. 

You said you removed all the placement from the CR? I could definitely see that causing an update to the pod specs. I'm wondering if that could be related to the destructive behavior as well.

**Comment by mmack** [comment]
By destruction i mean all deployments of all mons and osds are deleted. 
Also the configmaps. Alex has seen it yesterday in a screensharing session.

Am 10. August 2019 01:13:10 schrieb Travis Nielsen <notifications@github.com>:

**Comment by travisn** [comment]
I wonder if this could be related to the ownerReferences change, backported to v1.0.3 with [this change](https://github.com/rook/rook/commit/c7e0eecbede4d0032f4f55d1d68a8ea07d0daf99)

Thought it's still not clear what would trigger the resources to be garbage collected. The only side effects of that issue I've observed is that it fixed the garbage collection in OpenShift, but no change in behavior for K8s.

**Comment by travisn** [comment]
Ok, this issue was fixed by #3530, which wasn't backported. This issue will only be hit in clusters where the `name` of the CR is different from the `namespace` of the cluster. See #3530 for more explanation.

**Comment by christianhuening** [comment]
@travisn is there an ETA for `1.0.5`?

**Comment by travisn** [comment]
Due to the severity of this issue and that the fix has already been verified in master, I'll see if we can get this backported and even release v1.0.5 today.

**Comment by travisn** [comment]
Fixed in #3623, preparing the release now...

**Comment by travisn** [comment]
To be clear, this issue will only potentially hit clusters that have changed either the `name` or the `namespace` where the CephCluster CR is created. If they are both the same, the issue will not be hit. Since the default for both is `rook-ceph` (see the [default example](https://github.com/rook/rook/blob/release-1.0/cluster/examples/kubernetes/ceph/cluster.yaml#L15-L16)), most clusters will not hit this issue.
